### PR TITLE
Add direct stdin support to SafeCmd (#30)

### LINKED ISSUE #30: Add stdin/input_text support to SafeCmd.run/run_sync (or ExecutionContext)
## Summary

It would be valuable to pass raw stdin content directly into a `SafeCmd` execution (e.g., `input_text` or `input_bytes`) without constructing a pipeline or spawning helper commands like echo. This improves ergonomics and avoids expanding allowlists just to feed stdin.

## Motivation / Use Cases

- Running tools that expect stdin (e.g., `sed`, `jq`, `psql`, `git` hash-object --stdin`) in scripts or CI.
- Avoiding helper commands for stdin injection to keep allowlists small and reduce overhead.
- Consistency with `subprocess.run(input=...)` / `subprocess.run(text=True)` style usage.

### Option A — extend run/run_sync:

```python
result = cmd.run_sync(input_text="hello\n")
result = await cmd.run(input_bytes=b"...")
```

### Option B — add to ExecutionContext:

```python
ctx = ExecutionContext(stdin_text="hello\n")  # or stdin_bytes=b"..."
result = cmd.run_sync(context=ctx)
```

### Behavior

- `input_text` implies text encoding using `ExecutionContext.encoding`/`errors` (defaults already exist).
- `input_bytes` writes raw bytes.
- If both `input_text` and `input_bytes` are supplied, raise a `ValueError`.
- If capture is False, stdin injection should still work.
- This should not require an additional allowlisted program.

### Backwards Compatibility

- Default remains no stdin injection.
- New parameters are optional and non‑breaking.

### Suggested Tests

- Single command receives stdin and outputs expected stdout.
- Non‑zero exit code still captured correctly with stdin input.
- Conflicting parameters raise validation error.

## Why this matters

Current usage requires a pipeline `(echo(...) | cmd(...))`, which forces extra programs into the allowlist and increases
overhead/complexity. Direct stdin support keeps Cuprum’s allowlist model intact while improving usability.

### PR DESCRIPTION
**Pull request by leynos**
## Summary

This branch adds direct stdin input to `SafeCmd.run()` and `run_sync()` so callers can pass text or bytes without constructing helper pipelines or broadening allowlists.

Closes #30.

## Review walkthrough

- Start with [cuprum/sh.py](https://github.com/leynos/cuprum/blob/issue-30-add-stdin-input-text-support-to-safecmd/cuprum/sh.py) to review the new `input_text` and `input_bytes` API validation and encoding path.
- Then review [cuprum/_subprocess_execution.py](https://github.com/leynos/cuprum/blob/issue-30-add-stdin-input-text-support-to-safecmd/cuprum/_subprocess_execution.py) for subprocess stdin pipe creation, concurrent writing, and broken-pipe handling.
- Check [cuprum/unittests/test_safe_cmd_run.py](https://github.com/leynos/cuprum/blob/issue-30-add-stdin-input-text-support-to-safecmd/cuprum/unittests/test_safe_cmd_run.py) and [tests/behaviour/test_execution_runtime.py](https://github.com/leynos/cuprum/blob/issue-30-add-stdin-input-text-support-to-safecmd/tests/behaviour/test_execution_runtime.py) for coverage of text, bytes, encoding, validation, non-zero exits, disabled capture, and behavioural runtime usage.
- Finish with [docs/users-guide.md](https://github.com/leynos/cuprum/blob/issue-30-add-stdin-input-text-support-to-safecmd/docs/users-guide.md) and [docs/cuprum-design.md](https://github.com/leynos/cuprum/blob/issue-30-add-stdin-input-text-support-to-safecmd/docs/cuprum-design.md) for the user-facing and design documentation updates.

## Validation

- `make build`: passed
- `uv run pytest -q cuprum/unittests/test_safe_cmd_run.py tests/behaviour/test_execution_runtime.py`: 55 passed
- `make check-fmt`: passed
- `make lint`: passed
- `make typecheck`: passed
- `make test`: 486 passed, 43 skipped
- `make markdownlint`: passed
- `make nixie`: passed

## Notes

- `coderabbit review --agent` was attempted twice after the implementation milestone, but the service returned a recoverable rate-limit error before producing review findings.

## Summary by Sourcery

Add direct stdin input support to SafeCmd commands and wire it through the subprocess execution runtime with corresponding tests and documentation updates.

New Features:
- Allow SafeCmd.run() and run_sync() to accept input_text and input_bytes parameters for feeding data directly to subprocess stdin.

Enhancements:
- Extend the async subprocess execution layer to optionally pipe caller-provided stdin bytes, handling timeouts, cancellation, and processes that close stdin early.

Documentation:
- Update the user guide and design documentation to describe the new direct stdin input parameters and their interaction with execution context encodings.

Tests:
- Add unit and behavioural tests covering direct stdin text/bytes input, encoding behaviour, validation errors when both forms are supplied, interaction with capture settings, and non-zero exit handling.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[cuprum/_subprocess_execution.py](https://github.com/leynos/cuprum/pull/60/files/86436554abd3dad331c7be990b9f7e694d5f7484#diff-3974bb1efe0f6d53395be510d76c91deb50a8e830ac8b47047408d816fa382cc)

Comment on file

~~~python
        errors=execution.ctx.errors,
    )
    consumers = _spawn_stream_consumers(process, execution, stream_config, pid=pid)
    stdin_task = _spawn_stdin_writer(process, execution.stdin_data)
~~~

❌ New issue: [Complex Method](https://codescene.io/projects/74471/delta?repo-id=1354509&review-id=60&biomarker=Complex+Method&filename=cuprum%2F_subprocess_execution.py&method=_run_subprocess_with_streams)
_run_subprocess_with_streams has a cyclomatic complexity of 10, threshold = 9

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[cuprum/sh.py](https://github.com/leynos/cuprum/pull/60/files/86436554abd3dad331c7be990b9f7e694d5f7484#diff-8b408763a095cb52608b95f2dbe224ee541c0af855ea13cc60134a903859bb0a)

Comment on file

~~~python
        return Pipeline.concat(self, other)

    async def run(
    async def run(  # noqa: PLR0913 - mirrors subprocess.run-style runtime API
~~~

❌ New issue: [Complex Method](https://codescene.io/projects/74471/delta?repo-id=1354509&review-id=60&biomarker=Complex+Method&filename=cuprum%2Fsh.py&method=run)
run has a cyclomatic complexity of 10, threshold = 9

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[cuprum/sh.py](https://github.com/leynos/cuprum/pull/60/files/86436554abd3dad331c7be990b9f7e694d5f7484#diff-8b408763a095cb52608b95f2dbe224ee541c0af855ea13cc60134a903859bb0a)

Comment on file

~~~python
        return Pipeline.concat(self, other)

    async def run(
    async def run(  # noqa: PLR0913 - mirrors subprocess.run-style runtime API
~~~

❌ New issue: [Excess Number of Function Arguments](https://codescene.io/projects/74471/delta?repo-id=1354509&review-id=60&biomarker=Excess+Number+of+Function+Arguments&filename=cuprum%2Fsh.py&method=run)
run has 6 arguments, max arguments = 4

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[cuprum/sh.py](https://github.com/leynos/cuprum/pull/60/files/86436554abd3dad331c7be990b9f7e694d5f7484#diff-8b408763a095cb52608b95f2dbe224ee541c0af855ea13cc60134a903859bb0a)

Comment on file

~~~python
        return result

    def run_sync(
    def run_sync(  # noqa: PLR0913 - mirrors run() for sync callers
~~~

❌ New issue: [Excess Number of Function Arguments](https://codescene.io/projects/74471/delta?repo-id=1354509&review-id=60&biomarker=Excess+Number+of+Function+Arguments&filename=cuprum%2Fsh.py&method=run_sync)
run_sync has 6 arguments, max arguments = 4

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

## Complex Method
cuprum/sh.py: run

### What lead to degradation?
run has a cyclomatic complexity of 10, threshold = 9

### Why does this problem occur?
A Complex Method has a high cyclomatic complexity. The recommended threshold for the Python language is a cyclomatic complexity lower than 9.

### How to fix it?
There are many reasons for Complex Method. Sometimes, another design approach is beneficial such as a) modeling state using an explicit state machine rather than conditionals, or b) using table lookup rather than long chains of logic. In other scenarios, the function can be split using [EXTRACT FUNCTION](https://refactoring.com/catalog/extractFunction.html). Just make sure you extract natural and cohesive functions. Complex Methods can also be addressed by identifying complex conditional expressions and then using the [DECOMPOSE CONDITIONAL](https://refactoring.com/catalog/decomposeConditional.html) refactoring.

### Helpful refactoring examples
To get a general understanding of what this code health issue looks like - and how it might be addressed - we have prepared some diffs for illustrative purposes.

#### SAMPLE

~~~diff
# complex_method.js
 function postItem(item) {
   if (!item.id) {
-    if (item.x != null && item.y != null) {
-      post(item);
-    } else {
-      throw Error("Item must have x and y");
-    }
+    // extract a separate function for creating new item
+    postNew(item);
   } else {
-    if (item.x < 10 && item.y > 25) {
-      put(item);
-    } else {
-      throw Error("Item must have an x and y value between 10 and 25");
-    }
+    // and one for updating existing items
+    updateItem(item);
   }
 }
+
+function postNew(item) {
+  validateNew(item);
+  post(item);
+}
+
+function updateItem(item) {
+  validateUpdate(item);
+  put(item);
+}
+
~~~

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

## Excess Number of Function Arguments
cuprum/sh.py: run

### What lead to degradation?
run has 6 arguments, max arguments = 4

### Why does this problem occur?
Functions with many arguments indicate either a) low cohesion where the function has too many responsibilities, or b) a missing abstraction that encapsulates those arguments.
The threshold for the Python language is 4 function arguments.

### How to fix it?
Start by investigating the responsibilities of the function. Make sure it doesn't do too many things, in which case it should be split into smaller and more cohesive functions. Consider the refactoring [INTRODUCE PARAMETER OBJECT](https://refactoring.com/catalog/introduceParameterObject.html) to encapsulate arguments that refer to the same logical concept.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

## Excess Number of Function Arguments
cuprum/sh.py: run_sync

### What lead to degradation?
run_sync has 6 arguments, max arguments = 4

### Why does this problem occur?
Functions with many arguments indicate either a) low cohesion where the function has too many responsibilities, or b) a missing abstraction that encapsulates those arguments.
The threshold for the Python language is 4 function arguments.

### How to fix it?
Start by investigating the responsibilities of the function. Make sure it doesn't do too many things, in which case it should be split into smaller and more cohesive functions. Consider the refactoring [INTRODUCE PARAMETER OBJECT](https://refactoring.com/catalog/introduceParameterObject.html) to encapsulate arguments that refer to the same logical concept.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for the following concern and provide an AI coding agent prompt for the fix:

**Preserve inherited stdin when no direct input is provided**

Switching `stdin` to `asyncio.subprocess.DEVNULL` when `stdin_data` is absent changes existing command behavior: subprocesses no longer inherit the parent stdin and will see immediate EOF. This regresses common cases like running a SafeCmd from a process with piped stdin (or interactive tools expecting terminal input), which worked before because `create_subprocess_exec` inherited stdin by default. Use inherited stdin (`None`) when no explicit `StdinInput` is supplied, and only open a pipe when direct stdin data is requested.

**Comment by leynos** [comment]
@coderabbitai review

File: cuprum/_subprocess_execution.py [hunk 1]
```diff
@@ -99,11 +102,45 @@ async def _spawn_subprocess(
             if execution.capture or execution.echo
             else asyncio.subprocess.DEVNULL
         ),
+        stdin=(
+            asyncio.subprocess.PIPE
+            if execution.stdin_data is not None
+            else asyncio.subprocess.DEVNULL
```
[the remark this answers is not in the document]
Comment (leynos) [reply cuprum/_subprocess_execution.py]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (2 errors, 9 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Unit Architecture | ❌ Error | Four unaddressed review comments indicate incomplete refactoring: unused \_resolve\_stdin\_data, missing IOOptions deprecation, unhandled stdin\_task cancellation, unjustified lint suppression. | Delete \_resolve\_stdin\_data. Add deprecation warning to IOOptions.**post\_init**. Cancel stdin\_task before gathering on CancelledError. Justify file-level suppression. |
| Security And Privacy | ❌ Error | Unused \_resolve\_stdin\_data function; IOOptions lacks deprecation **post\_init**; stdin\_task not cancelled on CancelledError; file-level suppression lacks tracking comment. | Delete \_resolve\_stdin\_data; add deprecation warning to IOOptions.**post\_init**; call stdin\_task.cancel() before gather on CancelledError; justify too-many-lines suppression with ticket. |
| User-Facing Documentation | ⚠️ Warning | README Quick taste uses `io=IOOptions` but code requires `output=RunOutputOptions`. Users-guide prose references old API keywords, not `RunOutputOptions`. | Fix README to use `output=RunOutputOptions(echo=True)`. Update users-guide prose to reference the `output` parameter. Add migration guide for SafeCmd API change (capture/echo to RunOutputOptions). |
| Developer Documentation | ⚠️ Warning | Subprocess stdin architecture undocumented in developers-guide.md; roadmap missing item for stdin feature (issue `#30`). | Document subprocess execution internals and stdin architecture in developers-guide.md. Add completed roadmap item 1.2.3 for stdin support. |
| Testing (Unit And Behavioural) | ⚠️ Warning | Tests lack coverage for stdin interaction with timeouts and cancellation. The implementation includes stdin task cleanup during CancelledError but has no test exercising this critical path. | Add unit tests: test\_stdin\_input\_with\_timeout\_escalation and test\_stdin\_input\_cancellation\_cleans\_up\_task, exercising stdin provision during timeout/cancellation scenarios. |
| Testing (Property / Proof) | ⚠️ Warning | PR introduces encoding, concurrency, and state-transition invariants for stdin handling but lacks property-based tests despite Hypothesis precedent in codebase (test\_stream\_property\_based.py). | Add Hypothesis property tests for: (1) StdinInput encoding across Unicode/encodings; (2) stdin payload sizes/boundaries; (3) cancellation/timeout timing with stdin writes. |
| Domain Architecture | ⚠️ Warning | Four unresolved review issues: unused \_resolve\_stdin\_data, IOOptions missing deprecation warning, stdin\_task not cancelled on CancelledError, unresolved lint suppression. | Delete \_resolve\_stdin\_data; add DeprecationWarning to IOOptions; cancel stdin\_task on CancelledError; add tracking ticket for lint suppression. |
| Observability | ⚠️ Warning | Stdin write failures are silently suppressed without logging, metrics, or trace events, making production diagnostics impossible when processes close stdin early or write fails. | Add observability: emit 'stdin\_error' events in \_write\_stdin when BrokenPipeError/ConnectionResetError occur; document silent broken-pipe handling in users guide; optionally add debug-level logging for stdin write failures. |
| Performance And Resource Use | ⚠️ Warning | stdin\_task not cancelled before gather on CancelledError; unused \_resolve\_stdin\_data; missing DeprecationWarning in IOOptions; stdin data lacks size bounds. | Cancel stdin\_task before gather on CancelledError; remove \_resolve\_stdin\_data; add DeprecationWarning to IOOptions; document stdin size limits. |
| Concurrency And State | ⚠️ Warning | Stdin writer task can hang on CancelledError: code gathers stdin\_task without cancelling it first, risking deadlock if process exits while drain() is blocked on full pipe buffer. | Cancel stdin\_task explicitly before gathering: add `stdin_task.cancel()` before the `await asyncio.gather(stdin_task, ...)` call at line 254 in \_subprocess\_execution.py to ensure prompt cleanup on cancellation. |
| Architectural Complexity And Maintainability | ⚠️ Warning | Four review comments unaddressed: unused \_resolve\_stdin\_data duplicates logic; stdin\_task not cancelled on CancelledError; IOOptions lacks deprecation warning; pylint suppression lacks justification. | Delete \_resolve\_stdin\_data; cancel stdin\_task before gather on CancelledError; add DeprecationWarning to IOOptions; justify pylint suppression. |

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for the following concern and provide an AI coding agent prompt for the fix:

PR introduces encoding, concurrency, and state-transition invariants for stdin handling but lacks property-based tests despite Hypothesis precedent in codebase (test\_stream\_property\_based.py). Add Hypothesis property tests for: (1) StdinInput encoding across Unicode/encodings; (2) stdin payload sizes/boundaries; (3) cancellation/timeout timing with stdin writes.

Hypothesis is an approved dependency for df12 Productions Python code bases and property tests are expected as de rigueur where appropriate.

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (1 error, 5 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Unit Architecture | ❌ Error | StdinInput.resolve() can raise UnicodeEncodeError but this is undocumented in SafeCmd.run() Raises section, violating requirement that fallibility be explicit at operation boundaries. | Document UnicodeEncodeError in SafeCmd.run/run\_sync docstrings and StdinInput.resolve; add test exercising encoding failures. |
| Developer Documentation | ⚠️ Warning | Developers-guide omits \_execute\_with\_hooks and \_build\_stream\_config documentation. Roadmap missing entry for issue `#30` stdin support. | Add section 3.5 to roadmap for issue `#30`. Document \_execute\_with\_hooks and \_build\_stream\_config helper functions in developers-guide. |
| Testing (Compile-Time / Ui) | ⚠️ Warning | stdin\_error events are emitted by \_subprocess\_execution.py but never tested; snapshot tests should verify event emission when stdin writes fail. | Add snapshot test in test\_observe.py verifying stdin\_error events are emitted with correct phase and note fields when stdin write fails. |
| Observability | ⚠️ Warning | Stdin handling adds new async behaviour without adequate observability: no logging at decision points, no metrics for throughput/latency, and stdin\_error events lack test validation. | Add logging at stdin write paths; add test validating stdin\_error events are emitted; consider metrics for stdin throughput and document logging structure. |
| Concurrency And State | ⚠️ Warning | \_wait\_for\_exit\_code gathers stdin\_task without cancelling it on CancelledError, whilst \_run\_subprocess\_with\_streams explicitly cancels first, creating inconsistent cleanup. | Add stdin\_task.cancel() before gather in \_wait\_for\_exit\_code CancelledError handler to match streaming path and ensure consistent stdin writer cleanup. |
| Architectural Complexity And Maintainability | ⚠️ Warning | RunOutputOptions adds indirection without validation. SafeCmd.run breaks old API (capture/echo gone), contradicting backwards-compatibility. Pipeline.run keeps old API, causing inconsistency. | Restore SafeCmd.run(capture/echo) via overloads, or document breaking change. Align Pipeline.run to use RunOutputOptions or consolidate with ExecutionContext. |

**Comment by leynos** [comment]
@coderabbitai please create a new GitHub issue for the following:

### ⚠️ Architectural — `Pipeline.run` API inconsistency

The migration guide in `users-guide.md` is present and documents the `capture=`/`echo=` → `RunOutputOptions` transition for `SafeCmd`. That part is resolved. However, `Pipeline.run` and `Pipeline.run_sync` still expose flat `capture`/`echo` kwargs, creating a genuine API inconsistency. Aligning `Pipeline` would require updating its call sites throughout benchmarks, concurrent helpers, and tests — that is out of scope for this PR.

Shall I open a GitHub issue to track the `Pipeline.run` → `RunOutputOptions` alignment?

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (1 error, 5 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Unit Architecture | ❌ Error | StdinInput.resolve() can raise UnicodeEncodeError but this is undocumented in SafeCmd.run() Raises section, violating requirement that fallibility be explicit at operation boundaries. | Document UnicodeEncodeError in SafeCmd.run/run\_sync docstrings and StdinInput.resolve; add test exercising encoding failures. |
| Developer Documentation | ⚠️ Warning | Developers-guide omits \_execute\_with\_hooks and \_build\_stream\_config documentation. Roadmap missing entry for issue `#30` stdin support. | Add section 3.5 to roadmap for issue `#30`. Document \_execute\_with\_hooks and \_build\_stream\_config helper functions in developers-guide. |
| Testing (Compile-Time / Ui) | ⚠️ Warning | stdin\_error events are emitted by \_subprocess\_execution.py but never tested; snapshot tests should verify event emission when stdin writes fail. | Add snapshot test in test\_observe.py verifying stdin\_error events are emitted with correct phase and note fields when stdin write fails. |
| Observability | ⚠️ Warning | Stdin handling adds new async behaviour without adequate observability: no logging at decision points, no metrics for throughput/latency, and stdin\_error events lack test validation. | Add logging at stdin write paths; add test validating stdin\_error events are emitted; consider metrics for stdin throughput and document logging structure. |
| Concurrency And State | ⚠️ Warning | \_wait\_for\_exit\_code gathers stdin\_task without cancelling it on CancelledError, whilst \_run\_subprocess\_with\_streams explicitly cancels first, creating inconsistent cleanup. | Add stdin\_task.cancel() before gather in \_wait\_for\_exit\_code CancelledError handler to match streaming path and ensure consistent stdin writer cleanup. |
| Architectural Complexity And Maintainability | ⚠️ Warning | RunOutputOptions adds indirection without validation. SafeCmd.run breaks old API (capture/echo gone), contradicting backwards-compatibility. Pipeline.run keeps old API, causing inconsistency. | Restore SafeCmd.run(capture/echo) via overloads, or document breaking change. Align Pipeline.run to use RunOutputOptions or consolidate with ExecutionContext. |

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

[cuprum/unittests/test_observe.py](https://github.com/leynos/cuprum/pull/60/files/20ece21804137048a2892794c7c242db477758b3#diff-07690531ed48483510e58095f18114a1b09a02f58cd955e44047deb48184dddf)

Comment on lines +273 to +293

~~~python
def test_write_stdin_observes_runtime_error_from_wait_closed() -> None:
    """Emit stdin_error details when wait_closed fails."""
    stdin = _FailingStdin(wait_closed_error=RuntimeError("loop closed"))
    process = _FakeProcess(stdin)
    observation = _FakeObservation()

    asyncio.run(
        _write_stdin(
            typ.cast("asyncio.subprocess.Process", process),
            b"payload",
            typ.cast("_StageObservation", observation),
        )
    )

    assert stdin.closed is True
    assert observation.events == [
        (
            "stdin_error",
            _EventDetails(pid=123, note="RuntimeError: loop closed"),
        )
    ]
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/74471/delta?repo-id=1354509&review-id=60&biomarker=Code+Duplication&filename=cuprum%2Funittests%2Ftest_observe.py&method=)
The module contains 2 functions with similar structure: test_write_stdin_observes_os_error_from_drain,test_write_stdin_observes_runtime_error_from_wait_closed

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

#### cuprum/unittests/test_observe.py (1)

`235-246`: _⚠️ Potential issue_ | _🟠 Major_ | _⚡ Quick win_
**Make this stdin-error integration test deterministic.**
Stop relying on `262144` bytes to overflow the pipe. That assumption is not portable: pipe and transport buffers can exceed 256 KiB, so the write can complete without any `stdin_error` event and this assertion becomes flaky in CI. Drive the failure with a deterministic synchronisation point or a controlled stub instead of buffer-pressure heuristics.

##### 🤖 Detailed instructions

Verify each finding against current code. Fix only still-valid issues, skip the
rest with a brief reason, keep changes minimal, and validate.

In `@cuprum/unittests/test_observe.py` around lines 235 - 246, The test currently
relies on writing 262144 bytes (large_payload) to overflow the pipe which is
flaky; instead make the failure deterministic by forcing the subprocess to close
its stdin before the writer writes or by stubbing the writer to raise a
BrokenPipe: replace the large_payload heuristic with a small payload and
start/run a command via cmd.run_sync (or use the existing sh.observe(hook) hook)
that immediately exits or explicitly closes its stdin so the write fails, or
alternatively inject a controlled stub/mocked StdinInput that raises an EPIPE
when consumed; update the test that asserts on events (events and stdin_error)
to use this deterministic early-close or stub to reliably generate the
stdin_error event.

**Comment by leynos** [comment]
@coderabbitai review
